src/hpc/tls.h

```diff
@@ -55,19 +55,40 @@ typedef struct ThreadLocalStorage
   UInt LocksContended;
 } ThreadLocalStorage;
 
-#ifdef HAVE_NATIVE_TLS
+#ifdef USE_NATIVE_TLS
 extern __thread ThreadLocalStorage *TLSInstance;
 #endif
 
+#ifdef USE_PTHREAD_TLS
+
+#include <pthread.h>
+
+#ifdef __GNUC__
+// This trick allows gcc/clang to hoist calls to pthread_getspecific()
+// out of loops and perform constant subexpression elimination on
+// expressions containing TLS values.
+__attribute__((pure)) void * pthread_getspecific(pthread_key_t);
+#endif
+
+static pthread_key_t TLSKeyCopy;
+pthread_key_t GetTLSKey(void);
+
+__attribute__((constructor))
+static void InitTLSKey() {
+    TLSKeyCopy = GetTLSKey();
+}
+#endif
+
 #if defined(VERBOSE_GUARDS)
 static inline ThreadLocalStorage *GetTLS(void)
 #else
 static ALWAYS_INLINE ThreadLocalStorage *GetTLS(void)
 #endif
 {
-#ifdef HAVE_NATIVE_TLS
+#ifdef USE_NATIVE_TLS
     return TLSInstance;
 #else
+#ifndef USE_PTHREAD_TLS
```

Comment:
I'd prefer to use non-negated conditions; that's of course subjective, but I find it easier to understand the code flow here. Also, here we could actually avoid the nesting, and do this instead:
```C
#if defined(USE_NATIVE_TLS)
 ...
#elif defined(USE_MACOS_PTHREAD_TLS_ASM)
...
#elif defined(USE_PTHREAD_TLS)
...
#else
#error no support TLS method detected
#endif
```

Reply:
Restructured that, though the logic looks different. `USE_MACOS_PTHREAD_TLS_ASM` has to be nested under the `USE_PTHREAD_TLS` case.